hera_cal/apply_cal.py

```diff
@@ -246,20 +292,35 @@ def apply_cal(data_infilename, data_outfilename, new_calibration, old_calibratio
                 if flag_file is not None:
                     data_flags[bl] = np.logical_or(data_flags[bl], ext_flags[bl])
             if redundant_solution:
-                from .redcal import get_reds
-                all_reds = get_reds(hd.antpos, pols=hd.pols, bl_error_tol=bl_error_tol)
                 calibrate_redundant_solution(data, data_flags, new_gains, new_flags, all_reds, old_gains=old_gains,
                                              old_flags=old_flags, gain_convention=gain_convention)
             else:
                 calibrate_in_place(data, new_gains, data_flags=data_flags, cal_flags=new_flags,
                                    old_gains=old_gains, gain_convention=gain_convention)
-            hd.partial_write(data_outfilename, data=data, flags=data_flags,
-                             inplace=True, clobber=clobber, add_to_history=add_to_history)
-
+            hd.update(data=data, flags=data_flags)
+
+            if redundant_average:
+                # by default, weight by nsamples (but not flags). This prevents spectral structure from being introduced
+                # and also allows us to compute redundant averaged vis in flagged channels (in case flags are spurious).
+                if no_red_weights:
+                    redundant_weights = copy.deepcopy(data_nsamples)
+                    for bl in data_flags:
+                        if np.all(data_flags[bl]):
+                            redundant_weights[bl][:] = 0.
+                # redundantly average
+                utils.red_average(data=data, flags=data_flags, nsamples=data_nsamples, reds=all_red_antpairs, wgts=redundant_weights, inplace=True)
+                # update redundant data. Don't partial write.
+                hd_red.update(nsamples=data_nsamples, flags=data_flags, data=data)
+            else:
+                # partial write works for no redundant averaging.
+                hd.partial_write(data_outfilename, inplace=True, clobber=clobber, add_to_history=add_to_history, **kwargs)
+        if redundant_average:
+            # if we did redundant averaging, just write the redundant dataset out in the end at once.
+            hd_red.write_uvh5(data_outfilename, clobber=clobber)
```

Comment:
I think you should pass through ` **kwargs`

Reply:
added this.